### DISCUSSION TITLE
Problems with implementing MPC using Casadi Optistack

### DISCUSSION THREAD

**Discussion by Jushan-Randy-Chen** [casadi/casadi#3240] [Q&A]
Hi all,

I am running an MPC problem, but I've been creating new instances of `Opti( ) ` and all relevant constraints for every MPC iteration, and each successive iterates are related through the initial condition only (shifting horizons), which is not efficient at all. I wonder if there is a way to create a single `Opti ( )` instance only and then warm start all subsequent iterations? 

For example, at the beginning my first iteration is optimized over a horizon of `T=10`, and my second iteration is shifted 1 time step forward, which overlaps with the trajectory from the first iteration starting at the second time stamp. Therefore, due to the heavy overlapping, I don't think it makes sense to create new instances of symbolic variables for each iteration.

**Comment by brunoml5** [comment]
Hi!

You must use the solution from the last call as a warm start. Take a look at this [part of the documentation](https://web.casadi.org/docs/#problem-solving-and-retrieving).

Also, don't forget to use parameters in your optimization. For MPC, the initial state can be a parameter, something like `X(0) == X0`, where `X0` is a parameter which can change at each solver call.

**Reply by Jushan-Randy-Chen** [reply]
Hi,

Thanks for the response. I just tried using warm start as described in the document, and the computation does improve a lot. Could you elaborate on how I should use a parameter to "seed" the initial condition of each MPC iteration? Right now I have `Opti( ).subject_to(initial_cond` for every MPC iteration.
Moreover, I still see a significant pause between each MPC optimization because I had to re-initialize an `Opti( ) ` object at each iteration. Is there a way to get around that?

**Reply by brunoml5** [reply]
Hi!

If you have to create an Opti or solver object each time, you are doing something wrong.

Take a look at this modification of the race_car.py example:
```
from casadi import *

N = 100 # number of control intervals

opti = Opti() # Optimization problem

X0 = opti.parameter(2,1)    # parameter variable for x0

X = opti.variable(2,N+1)    # state trajectory
pos   = X[0,:]
speed = X[1,:]
U = opti.variable(1,N)      # control trajectory (throttle)
T = opti.variable()         # final time

opti.minimize(T)    # race in minimal time

f = lambda x,u: vertcat(x[1],u-x[1])    # dx/dt = f(x,u)

dt = T/N    # length of a control interval
for k in range(N):  # loop over control intervals
   k1 = f(X[:,k],         U[:,k])
   k2 = f(X[:,k]+dt/2*k1, U[:,k])
   k3 = f(X[:,k]+dt/2*k2, U[:,k])
   k4 = f(X[:,k]+dt*k3,   U[:,k])
   x_next = X[:,k] + dt/6*(k1+2*k2+2*k3+k4)

   opti.subject_to(X[:,k+1]==x_next)    # close the gaps

limit = lambda pos: 1-sin(2*pi*pos)/2
opti.subject_to(speed<=limit(pos))      # track speed limit
opti.subject_to(opti.bounded(0,U,1))    # control is limited

opti.subject_to(pos[-1]==1)  # finish line at position 1

opti.subject_to(T>=0) # Time must be positive
```

Notice the following constraint which makes X(0) == X0
```
opti.subject_to(pos[0]==X0[0])      # Constraints for X0
opti.subject_to(speed[0]==X0[1])

opti.set_initial(speed, 1)  # initial values for solver
opti.set_initial(T, 1)

opti.solver("ipopt")    # Create solver object

opti.set_value(X0, vertcat([0, 0]))     # Set value for X0 before calling the solver
sol1 = opti.solve()   # actual solve
```

For the second solve, don't forget to set the new value for X0.
```
opti.set_initial(sol1.value_variables())    # warm start the decision variables

opti.set_value(X0, vertcat([0.0002, 0.02])) # Change the X0 parameter
sol2 = opti.solve()                         # Solve again
```

Then, some plotting
```
from pylab import plot, step, figure, legend, show, spy

plot(sol1.value(speed),label="speed")
plot(sol1.value(pos),label="pos")
plot(limit(sol1.value(pos)),'r--',label="speed limit")
step(range(N),sol1.value(U),'k',label="throttle")
legend(loc="upper left")

figure()
plot(sol2.value(speed),label="speed")
plot(sol2.value(pos),label="pos")
plot(limit(sol2.value(pos)),'r--',label="speed limit")
step(range(N),sol2.value(U),'k',label="throttle")
legend(loc="upper left")

show()
```

I hope it helps!

**Reply by Jushan-Randy-Chen** [reply]
Thanks! That helps. But when I run this MPC problem in a receding-horizon fashion, how should I modify the code above? Should I dump all the constraints associated with the `Opti( ) ` object at every MPC time step?
